segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1435 @@
+
+    /**
+     * {@link SystemJournal} that logs all changes to system segment layout so that they can be are used during system bootstrap.
+     */
+    @Getter
+    private SystemJournal systemJournal;
+
+    /**
+     * {@link ReadIndexCache} that has index of chunks by start offset
+     */
+    private final ReadIndexCache readIndexCache;
+
+    /**
+     * List of garbage chunks.
+     */
+    private final List<String> garbageChunks = new ArrayList<String>();
+
+    /**
+     * Prefix string to use for logging.
+     */
+    private String logPrefix;
+
+    /**
+     * Creates a new instance of the ChunkStorageManager class.
+     *
+     * @param chunkStorage ChunkStorageProvider instance.
+     * @param executor     An Executor for async operations.
+     * @param config       Configuration options for this ChunkStorageManager instance.
+     */
+    public ChunkStorageManager(ChunkStorageProvider chunkStorage, Executor executor, ChunkStorageManagerConfig config) {
+        this.config = Preconditions.checkNotNull(config, "config");
+        this.chunkStorage = Preconditions.checkNotNull(chunkStorage, "chunkStorage");
+        this.executor = Preconditions.checkNotNull(executor, "executor");
+        this.readIndexCache = new ReadIndexCache(config.getMaxIndexedSegments(),
+                config.getMaxIndexedChunksPerSegment(),
+                config.getMaxIndexedChunks());
+        this.closed = new AtomicBoolean(false);
+    }
+
+    /**
+     * Creates a new instance of the ChunkStorageManager class.
+     *
+     * @param chunkStorage  ChunkStorageProvider instance.
+     * @param metadataStore Metadata store.
+     * @param executor      An Executor for async operations.
+     * @param config        Configuration options for this ChunkStorageManager instance.
+     */
+    public ChunkStorageManager(ChunkStorageProvider chunkStorage, ChunkMetadataStore metadataStore, Executor executor, ChunkStorageManagerConfig config) {
+        this.config = Preconditions.checkNotNull(config, "config");
+        this.chunkStorage = Preconditions.checkNotNull(chunkStorage, "chunkStorage");
+        this.metadataStore = Preconditions.checkNotNull(metadataStore, "metadataStore");
+        this.executor = Preconditions.checkNotNull(executor, "executor");
+        this.readIndexCache = new ReadIndexCache(config.getMaxIndexedSegments(),
+                config.getMaxIndexedChunksPerSegment(),
+                config.getMaxIndexedChunks());
+        this.closed = new AtomicBoolean(false);
+    }
+
+    /**
+     * Initializes the ChunkStorageManager.
+     *
+     * @param metadataStore Metadata store.
+     * @param containerId   container id.
+     * @param systemJournal SystemJournal that keeps track of changes to system segments and helps with bootstrap.
+     * @throws Exception In case of any errors.
+     */
+    public void initialize(int containerId, ChunkMetadataStore metadataStore, SystemJournal systemJournal) throws Exception {
```

Comment:
Now there is a circular dependency between table store and storage. I need to plug into existing initialization logic in multiple steps. 
This is used like below in 3rd PR. 
https://github.com/sachin-j-joshi/pravega/commit/1cc42a43dbd9c16def003f227d1791b1a5b7bfbd#diff-b32d413ea6aa7bb2017179c6a1ca4511R189-R208

https://github.com/sachin-j-joshi/pravega/commit/1cc42a43dbd9c16def003f227d1791b1a5b7bfbd#diff-b32d413ea6aa7bb2017179c6a1ca4511R189-R208

https://github.com/sachin-j-joshi/pravega/commit/1cc42a43dbd9c16def003f227d1791b1a5b7bfbd#diff-b32d413ea6aa7bb2017179c6a1ca4511R189-R208

Reply:
The current design splits metadata management (ChunkMetadataStore) from the actual data path (ChunkStorageProvider).

- The specific metadata store can be any KV store that implements version based concurrency control and conditional updates based on version. (Most KV stores do.)
(In theory fencing can be implemented by keeping an epoch KV pair)
- My preference is that we build KV store that directly uses ChunkStorageProvider to store its pages.

- However we chose to use internal Table segments to store this metadata. Mainly because it is already implemented and available. Additionally we have BK fencing built into the implementation. 

But that also has a major down side - the table segment itself writes and reads data using `StorageWriter `- and already assumes that segment abstraction is provided by `Storage` interface. 

When table segment needs to read/write its data to underlying segments, it will end up calling into ChunkStorageManager.  But then ChunkStorageManager itself needs  to read/write all  KV metadata pairs to table segments.

All metadata about all segments and chunks is stored in table store by ChunkStorageManager - which is all fine , except what about those segments used by table segment itself?.
I call these special table segments as "storage system segments" or "critical segments" . Without information about these segments initialized, the table segments can't read or write , and without table segments we can't read/write metadata about other segments.

Those table segments themselves need to be stored in chunks and then metadata data about those segments/chunks needs to be managed.

So that is the origin of circular dependency. 

We break this dependency in following way.
1. The metadata records corresponding to "storage system segments" is pinned to memory and not written to table segments themselves. (marking record pinned means that KV pair is never evicted and for those records we don't look them up in underlying table segments)
2. Instead changes to these metadata is logged into `SystemJournal`. The system journal is written directly to underlying ChunkStorageProvider.

3. During initialization , we replay the SystemJournal log to recreate the state of  segments that make the table segments to store metadata.

With this background, during segment container initialization the storage has to be initialized before  table store extension , but then table segments need to be initialized and passed back to storage. So in the end everything works.

The obvious question at this point is that - should Storage have a dependency on table segments? 
And in retrospect I feel like that definitely complicates things.

The metadata can either be stored externally in some other store or locally in ChunkStorageProvider itself.

Alternative to using full table segment functionality is to implement Log Structured  Merge Tree (LSMT) based KV store on top of ChunkStorageProvider and also deal with fencing there.

Reply:
One thing I do not like here is that we pass yet another metadata store instance via this method, after we pass one to the constructor. Can we refactor this so that both the metadata store and the container id are passed at construction time? 

Another thing is that you have two initialized methods. You did explain below why, but this is utterly confusing to whomever is calling this. This particular initialize method should be put entirely within the constructor; feel free to make use of factories if that will help with things.

Then when the other initialize method is called, you can call whatever method is required on these classes to initialize themselves.

Reply:
| The metadata can either be stored externally in some other store or locally in ChunkStorageProvider itself.

We do not want to store metadata externally. That means we'd have to deploy yet another service with Pravega, which will make things more complicated. If you want to store it locally, how would do you do that?  You mentioned LSM Trees. Do you believe that doing something like that will be simpler than using this method? If you do it, you'd have to reimplement your fencing mechanism (no BK, so it doesn't come for free), deal with LTS problems, and on top of that, implement your own LTS layout. So you'd just be deferring the problem to another one.

Reply:
I don't understand why we had an initialization method on Storage interface ( ReadOnlyStorage) . But it is what it is - this interface us used all over the place. I can not change this at this point - Also I don't want to change all existing storage implementations everywhere at this point. 

I have already explained why table segment can not be passed in constructor in real container.

There are two constructor overloads. On of the overload is used in segment container class. The other one is used in lots of tests. 

However I can definitely rename the second initialize() method. 
```
/**
 * Defines a Read-Only abstraction for Permanent Storage with async operations.
 */
public interface ReadOnlyStorage extends AutoCloseable {
    /**
     * Initializes this Storage instance with the given ContainerEpoch.
     *
     * @param containerEpoch The Container Epoch to initialize with.
     */
    void initialize(long containerEpoch);
```

Reply:
Based on our conversation, resolving the discussion.